File: src/102. Triangle containment/102.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"math"
	"os"
	"strconv"
	"strings"
)

type Point struct {
	X, Y float64
}

type Triangle struct {
	A, B, C Point
}

func main() {
	origin := Point{0, 0}
	triangles := readTrianglesFromFile("p102_triangles.txt")
	count := 0
	for _, tri := range triangles {
		if isPointInsideTriangle(tri, origin) {
			count++
		}
	}
	fmt.Println(count)
}

func isPointInsideTriangle(tri Triangle, p Point) bool {
	areaABC := triangleArea(tri)
	a, b, c := tri.A, tri.B, tri.C
	areaABP := triangleArea(Triangle{a, b, p})
	areaBCP := triangleArea(Triangle{b, c, p})
	areaCAP := triangleArea(Triangle{c, a, p})
	return areaABC == areaABP + areaBCP + areaCAP
}

// https://en.wikipedia.org/wiki/Triangle#Using_coordinates
func triangleArea(tri Triangle) float64 {
	return math.Abs((tri.A.X * tri.B.Y - tri.A.X * tri.C.Y +
		tri.B.X * tri.C.Y - tri.B.X * tri.A.Y +
		tri.C.X * tri.A.Y - tri.C.X * tri.B.Y) / 2)
}

func readTrianglesFromFile(fileName string) []Triangle {
	f, err := os.Open(fileName)
	if err != nil {
		log.Fatal("err")
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	var triangles []Triangle
    for scanner.Scan() {
		coordinates := strings.Split(scanner.Text(), ",")
		var tri Triangle
		tri.A = buildPoint(coordinates[0:2])
		tri.B = buildPoint(coordinates[2:4])
		tri.C = buildPoint(coordinates[4:6])
		triangles = append(triangles, tri)
	}
	return triangles
}

func buildPoint(coordinates []string) Point {
	coordX, _ := strconv.ParseFloat(coordinates[0], 64)
	coordY, _ := strconv.ParseFloat(coordinates[1], 64)
	return Point{coordX, coordY}
}

File: src/102. Triangle containment/102.cpp
#include <cmath>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using std::string;
using std::vector;

struct Point {
    double x;
    double y;
};

struct Triangle {
    Point a;
    Point b;
    Point c;
};

static bool is_point_inside_trisiangle(const Triangle& tri, const Point& p);
static double triangle_area(const Triangle& tri);
static vector<Triangle> read_triangles_from_file(const char *file_name);
static Point build_point(const string& coord_x, const string& coord_y);

int main() {
    const Point origin = {0, 0};
    vector<Triangle> triangles = read_triangles_from_file("p102_triangles.txt");
    int count = 0;
    for (Triangle tri : triangles) {
        if (is_point_inside_trisiangle(tri, origin)) {
            count++;
        }
    }
    std::cout << count << std::endl;
}

static bool is_point_inside_trisiangle(const Triangle& tri, const Point& p) {
    double areaABC = triangle_area(tri);
    Point a = tri.a;
    Point b = tri.b;
    Point c = tri.c;
    double areaABP = triangle_area(Triangle{a, b, p});
    double areaBCP = triangle_area(Triangle{b, c, p});
    double areaCAP = triangle_area(Triangle{c, a, p});
    return areaABC == areaABP + areaBCP + areaCAP;
}

// https://en.wikipedia.org/wiki/Triangle#Using_coordinates
static double triangle_area(const Triangle& tri) {
    return std::abs((tri.a.x * tri.b.y - tri.a.x * tri.c.y +
        tri.b.x * tri.c.y - tri.b.x * tri.a.y +
        tri.c.x * tri.a.y - tri.c.x * tri.b.y) / 2);
}

static vector<Triangle> read_triangles_from_file(const char *file_name) {
    std::ifstream file_stream(file_name);
    string line;
    vector<Triangle> triangles;
    while (std::getline(file_stream, line)) {
        std::stringstream line_stream(line);
        vector<string> coordinates;
        string token;
        while (std::getline(line_stream, token, ',')) {
            coordinates.push_back(token);
        }
        Triangle tri;
        tri.a = build_point(coordinates[0], coordinates[1]);
        tri.b = build_point(coordinates[2], coordinates[3]);
        tri.c = build_point(coordinates[4], coordinates[5]);
        triangles.push_back(tri);
    }
    return triangles;
}

static Point build_point(const string& coord_x, const string& coord_y) {
    return Point{std::atof(coord_x.c_str()), std::atof(coord_y.c_str())};
}
